### DISCUSSION TITLE
Can't use server actions if file is marked as "use server"

### DISCUSSION THREAD

**Discussion by PabloGracia** [vercel/next.js#50976] [App Router] [answered]
I've recently started using the app router and the experimental server actions, and I'm not sure if I've found a bug or a feature, so I thought this would be the best way to look for answers.

My understanding is that with the app router, by default all components are server components **unless marked as client component**. That is, if I type "use client" in the top of the file, it will be a client component; if I don't type anything, it will be a server component. I've been marking my server components as "use server", to clearly state to myself that they are server components, as well as being able to import server actions into client components when those files include async actions.

However, I've found an issue with this approach when leveraging server actions. In a freshly created directory - leveraging create-next-app@latest - I have edited the main page to look like this:

```tsx
// app/page.tsx
"use server";

import fs from "fs/promises";

export default async function Home() {
  async function writeFile() {
    "use server";

    const data = "Hello, World!";
    const path = "public/hello.txt";
    const done = await fs.writeFile(path, data);

    return done;
  }

  return (
    <div className="flex min-h-screen flex-col items-center justify-between p-24">
      <h1>Welcome!</h1>
      <form action={writeFile}>
        <button type="submit">Write file</button>
      </form>
    </div>
  );
}
```

When running this code, I get the following error:
```
- error node_modules/next/dist/compiled/react-server-dom-webpack-experimental/cjs/react-server-dom-webpack-server.edge.development.js (1912:12) @ resolveModelToJSON
- error Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".
  <form action={function} children=...>
               ^^^^^^^^^^
    at stringify (<anonymous>)
null
```

Which is odd, as I haven't added any client components to my project. However, when I remove the `use server` directive from the top of the file, the code runs flawlessly, and the server action is executed upon clicking on the button.

What am I missing? As far as I understand (or assumed), there shouldn't be any difference between declaring "use server" or not on the top of this file, yet one of them fails stating that I cannot pass functions to a client component, whereas the other one doesn't.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

There's no `use client` counter part called `use server`, rather `use server` is used, as of now, to mark actions exclusively. 

So `use server` should be used in a file, with nothing but server actions.

- https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#with-client-components

> Note: When using a top-level "use server" directive, all exports below will be considered Server Actions. You can have multiple Server Actions in a single file.

So, as of now, I guess that makes the Home component a server action itself. At least that's my take on it, this is all alpha technology, so I wouldn't linger on it too much. Do as the docs says, and move on.

**Comment by JustinTBrown** [comment]
Just in case this helps anyone else getting the same error, but for a different reason: if you're importing a server action function from another file (eg. `utils/actions`), make sure the file is correctly named and includes the file extension. I accidentally omitted `.ts`. Once I corrected the filename (`utils/actions.ts`) the error cleared.

**Comment by oasaleh** [comment]
I don't get why do we need 'use server' if this is the default.

**Reply by icyJoseph** [reply]
> I don't get why do we need 'use server' if this is the default.

Maybe think of it in this scenario, which is not the only scenario possible 

- everything is a server component, but
- you eventually reach a part of the tree where, you need a client component
- this client component needs to trigger some server side work
- you mark the function/module as, use server, to tell the bundler, that even though you are within a client module, you want to keep this functions code in the server

> 
"use server" ensures this function is only ever executed on the server.

> 'use server' marks server-side functions that can be called from client-side code.

You can also ask yourself why it needs to be asynchronous? That's because these functions are designed to be called by external parties, without exposing their implementation, using a REST POST call.

So even when you have a React tree with nothing but Server components, that is zero client modules are made, you can still tell React to associate a form action with a piece of server side code, using a Server Action.

And you communicate this to the framework, though the use server directive

**Reply by oasaleh** [reply]
Thank you for clarifying. I have a follow up question, what's the difference between a server action (which is a function) and a UI component, which is also a function? Because according to Next's docs, you cannot have a server component under a client component in the tree. Or is it because I'm passing the server action as props (which is the only pattern you can do to render a server component under a client one)?

**Reply by icyJoseph** [reply]
>  Because according to Next's docs, you cannot have a server component under a client component in the tree.

You can have a Client Component which leaves a `slot`/`placeholder`, often `children`, and React renders the client component, around the slot. 

Like:

```jsx
// we are in a server component and we do...

<RevealsContentOnHover>
   <GetsDataFromMyDBServerSide />
</RevealsContentOnHover>
```

In this case, `RevealsContentOnHover`, perhaps is made in such a way that uses state, and onMouseEnter callbacks, etc, while, the nested component, `GetsDataFromMyDBServerSide` is a server component that gets data from a DB.

This is all possible, as long as it is all happening within a Server Component.

You wouldn't be able to do this if we were inside `RevealsContentOnHover`:

```jsx
// RevealsContentOnHover
'use client';

// later on

// ❌❌❌
<div onMouseEnter={() => {/* */}}>
  <GetsDataFromMyDBServerSide /> // This server component can't be owned by a `use client` file
</div>
```

What you can't really do is pass props into a server component, from a `use client` module. It might look like it works, but as soon as state changes, things break. Though I don't think it is fundamentally impossible, but at the moment, let's not go there.

>  what's the difference between a server action (which is a function) and a UI component, which is also a function?

Well, a server action is code that, using various techniques, can be "invoked" from an external system, namely the browser.

A UI component, is roughly a function... not gonna get into that too deep, but you don't ever do `App()`, right? You do `<App />`, and that's because JSX is make-belief/syntax-sugar for calls to, a so-called pragma, `React.createElement`, and since React 17, possibly other pragma, which at the end of the day, create React Elements, VDOM, or a plain object representation of a piece of UI. A function that takes in props, and returns a React Element, is a React component.

In the case of Server Components, what you get is a form of serialized VDOM. 

The React components you define, whether they are Server or Client components, are simply used, roughly speaking, to generate HTML, either at build time (pre-rendering), or as requests arrive (on-demand), and during hydration and render-passes on the client side.

**Reply by oasaleh** [reply]
It all makes so much sense now. Thank you so much!
